openvdb/tools/FindActiveValues.h

```diff
@@ -0,0 +1,410 @@
+        }
+    }
+    return count;
+}
+
+template<typename TreeT>
+template<typename NodeT>
+typename NodeT::NodeMaskType FindActiveValues<TreeT>::getBBoxMask(const CoordBBox &bbox, const NodeT* node) const
+{
+    typename NodeT::NodeMaskType mask;
+    auto b = node->getNodeBoundingBox();
+    assert( bbox.hasOverlap(b) );
+    if ( bbox.isInside(b) ) {
+        mask.setOn();//node is completely inside the bbox so early out
+    } else {
+        b.intersect(bbox);
+        b.min() &=  NodeT::DIM-1u;
+        b.min() >>= NodeT::ChildNodeType::TOTAL;
+        b.max() &=  NodeT::DIM-1u;
+        b.max() >>= NodeT::ChildNodeType::TOTAL;
+        assert( b.hasVolume() );
+        auto it = b.begin();
+        for (const Coord& x = *it; it; ++it) {
+            mask.setOn(x[2] + (x[1] << NodeT::LOG2DIM) + (x[0] << 2*NodeT::LOG2DIM));
+        }
+    }
+    return mask;
+}
+
+template<typename TreeT>
+template<typename NodeT>
+bool FindActiveValues<TreeT>::any(const NodeT* node, const CoordBBox &bbox) const
+{
+    // Generate a bit mask of the bbox coverage
+    auto mask = this->getBBoxMask(bbox, node);
+
+    // Check active tiles
+    const auto tmp = mask & node->getValueMask();// prune active the tile mask with the bbox mask
+    if (!tmp.isOff()) return true;
+
+    // Check child nodes
+    mask &= node->getChildMask();// prune the child mask with the bbox mask
+    const auto* table = node->getTable();
+    bool test = false;
+    for (auto i = mask.beginOn(); !test && i; ++i) {
+        test = this->any(table[i.pos()].getChild(), bbox);
+    }
+    return test;
+}
+
+template<typename TreeT>
+inline bool FindActiveValues<TreeT>::any(const typename TreeT::LeafNodeType* leaf, const CoordBBox &bbox ) const
+{
+    bool test = leaf->getValueMask().isOn();
+
+    for (auto i = leaf->cbeginValueOn(); !test && i; ++i) {
+        test = bbox.isInside(i.getCoord());
+    }
+    return test;
+}
+
+template<typename TreeT>
+inline Index64 FindActiveValues<TreeT>::count(const typename TreeT::LeafNodeType* leaf, const CoordBBox &bbox ) const
+{
+    Index64 count = 0;
+    if (leaf->getValueMask().isOn()) {
+        auto b = leaf->getNodeBoundingBox();
+        b.intersect(bbox);
+        count = b.volume();
+    } else {
+        for (auto i = leaf->cbeginValueOn(); i; ++i) {
+            if (bbox.isInside(i.getCoord())) ++count;
+        }
+    }
+    return count;
+}
+
+template<typename TreeT>
+template<typename NodeT>
+Index64 FindActiveValues<TreeT>::count(const NodeT* node, const CoordBBox &bbox) const
+{
+    Index64 count = 0;
+    // Generate a bit mask of the bbox coverage
+    auto mask = this->getBBoxMask(bbox, node);
+
+    // Check child nodes
+    const auto tmp = mask & node->getChildMask();// prune the child mask with the bbox mask
+    const auto* table = node->getTable();
+    for (auto i = tmp.beginOn(); i; ++i) {
+        count += this->count(table[i.pos()].getChild(), bbox);
+    }
```

Comment:
No multi-threading? This is probably the place where it would be most beneficial to use a parallel reduction?

Reply:
yeah I thought about it but I couldn't convince myself that the overhead of instantiating multiple threads per leaf node (of which there are of course many) would actually lead to a performance improvements. I guess I should at least try it (stay tuned for more :)

Reply:
added multi-threading at the internal node level. It did produce modest improvements on my laptop but it would be good it you could run it on a workstation with more threads.

Reply:
Sure, I can give it a go tomorrow. Did you consider using a NodeManager::reduceBottomUp() ? I figure that would be more efficient because it will parallelize across all leaf nodes / internal nodes in one go. Might not be worth the effort changing it though.